src/Toml/BiMap.hs

```diff
@@ -138,158 +170,169 @@ _Just = prism Just id
-_Day :: BiMap Day AnyValue
+_Day :: TomlBiMap Day AnyValue
 _Day = mkAnyValueBiMap (matchDate >=> getTime) (Date . Day)
   where
     getTime (Day d) = Just d
     getTime _       = Nothing
 
 -- | Time of day bimap for 'AnyValue'. Usually used with 'timeOfDay' combinator.
-_TimeOfDay :: BiMap TimeOfDay AnyValue
+_TimeOfDay :: TomlBiMap TimeOfDay AnyValue
 _TimeOfDay = mkAnyValueBiMap (matchDate >=> getTime) (Date . Hours)
   where
     getTime (Hours h) = Just h
     getTime _         = Nothing
 
 -- | Helper bimap for 'String' and 'Text'.
-_StringText :: BiMap String Text
+_StringText :: BiMap e String Text
 _StringText = iso T.pack T.unpack
 
 -- | 'String' bimap for 'AnyValue'. Usually used with 'string' combinator.
-_String :: BiMap String AnyValue
+_String :: TomlBiMap String AnyValue
 _String = _StringText >>> _Text
 
 -- | Helper bimap for 'String' and types with 'Read' and 'Show' instances.
-_ReadString :: (Show a, Read a) => BiMap a String
-_ReadString = BiMap (Just . show) readMaybe
+_ReadString :: (Show a, Read a) => TomlBiMap a String
+_ReadString = BiMap (Right . show) (eitherToBiEither . readEither)
 
 -- | Bimap for 'AnyValue' and values with a `Read` and `Show` instance.
 -- Usually used with 'read' combinator.
-_Read :: (Show a, Read a) => BiMap a AnyValue
+_Read :: (Show a, Read a) => TomlBiMap a AnyValue
 _Read = _ReadString >>> _String
 
 -- | Helper bimap for 'Natural' and 'Integer'.
-_NaturalInteger :: BiMap Natural Integer
-_NaturalInteger = BiMap (Just . toInteger) maybeInteger
+_NaturalInteger :: TomlBiMap Natural Integer
+_NaturalInteger = BiMap (Right . toInteger) eitherInteger
   where
-    maybeInteger :: Integer -> Maybe Natural
-    maybeInteger n
-      | n < 0     = Nothing
-      | otherwise = Just (fromIntegral n)
+    eitherInteger :: Integer -> Either TomlBiMapError Natural
+    eitherInteger n
+      | n < 0     = Left $ ArbitraryError "Value is below zero."
+      | otherwise = Right (fromIntegral n)
 
 -- | 'Natural' bimap for 'AnyValue'. Usually used with 'natural' combinator.
-_Natural :: BiMap Natural AnyValue
+_Natural :: TomlBiMap Natural AnyValue
 _Natural = _NaturalInteger >>> _Integer
 
 -- | Helper bimap for 'Integer' and integral, bounded values.
-_BoundedInteger :: (Integral a, Bounded a) => BiMap a Integer
-_BoundedInteger = BiMap (Just . toInteger) maybeBounded
+_BoundedInteger :: (Integral a, Bounded a) => TomlBiMap a Integer
+_BoundedInteger = BiMap (Right . toInteger) eitherBounded
   where
-    maybeBounded :: forall a. (Integral a, Bounded a) => Integer -> Maybe a
-    maybeBounded n
-      | n < toInteger (minBound :: a) = Nothing
-      | n > toInteger (maxBound :: a) = Nothing
-      | otherwise                     = Just (fromIntegral n)
+    eitherBounded :: forall a. (Integral a, Bounded a) => Integer -> Either TomlBiMapError a
+    eitherBounded n
+      | n < toInteger (minBound :: a) = Left $ ArbitraryError "Value is less minBound"
```

Comment:
It's better to patch error message here to tell what is the actual value and what is the value of `minBound`